### PR TITLE
gh-127111: Emscripten Make web example work again

### PR DESCRIPTION
**Pull request by hoodmane**
I moved the web example from `Tools/wasm` into `Tools/wasm/emscripten/web_example`. I also added a new target `build_emscripten` which is `build_wasm` but also builds the web_example. The web_example needs:
1. python.html, copied
2. python.worker.mjs copied
3. python.mjs and python.wasm output from the main linking of the Python interpreter
4. The webserver that sets COOP and COEP
5. python3.14.zip

This last is created by the `wasm_assets.py` script, which required a pretty small set of changes to work fine for us.

The last thing that should be done is the `python.worker.mjs` script should be made independent of the Python version: currently 3.14 is hard coded. [I ran into trouble doing this](https://github.com/emscripten-core/emscripten/issues/22980), so maybe I can leave it to a followup.

* Issue: gh-127111

### CODE REVIEW

File: Tools/wasm/emscripten/web_example/server.py
Comment (hoodmane) [inline Tools/wasm/emscripten/web_example/server.py:22]: This is now the default behavior of `SimpleHTTPRequestHandler`.

File: configure.ac
Comment (hoodmane) [inline configure.ac:2328]: This setting has changed names. It's not really relevant to this PR, I could separate it into a separate NFC commit.

File: configure.ac
Comment (hoodmane) [inline configure.ac:2335]: We use `callMain` in `python.worker.mjs` now so we need to ask for it to be exported.

File: configure.ac
Comment (hoodmane) [inline configure.ac:2342]: We also set this on line 2328. It's harmless but unnecessary to set it twice.

**Comment by hoodmane** [comment]
Thanks for testing it @JeffersGlass. Maybe I'll have to try a different CDN. It worked fine for me though...

File: configure.ac
Comment (freakboy3742) [reply configure.ac:2328]: I'm fine with including this change as part of this PR. We're still in a "get it working" stage, so we don't gain any real improvement by breaking it out.

File: Tools/wasm/emscripten/web_example/python.worker.mjs
Comment (freakboy3742) [inline Tools/wasm/emscripten/web_example/python.worker.mjs]: Unless it's going to be a major impediment to progress, I think we're going to need a solution for this before we can merge this PR - even though it's an example, it's code that will easily get forgotten along the way.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
The general shape of this makes sense. 

From a general process perspective, this needs a news entry to describe that that web example has been updated, and documentation in the WASM README on how to run the example.

However, assuming the usage instructions are `python Tools/wasm/emscripten build` and open `web_example/python.html` - I get 2 errors.
 
First, during the build process, the macOS version of `realpath` apparently doesn't honor the `-s` flag, so the usage on L223 of the `Tools/emscripten/__main__.py` raises an error and `python Tools/wasm/emscripten build` doesn't complete.

~Second, I wasn't able to see the example working in practice. When I open `file:///.../web_example/python.html` in Chrome, I get the following error:~
```
python.html:50 Uncaught (in promise) SecurityError: Failed to construct 'Worker': Script at 'file:///Users/rkm/projects/python/cpython/cross-build/wasm32-emscripten/web_example/python.worker.mjs' cannot be accessed from origin 'null'.
    at WorkerManager.initialiseWorker (python.html:50:27)
    at new WorkerManager (python.html:45:14)
    at window.onload (python.html:350:33)
```
~If I use `python -m http.server` to serve the `web_example` folder, and open python.html, I get:~
```
Uncaught ReferenceError: SharedArrayBuffer is not defined
    at new StdinBuffer (python.worker.mjs:5:24)
    at python.worker.mjs:62:21
```

EDIT: Strike that - I see from @JeffersGlass comment that `python web_example/server.py` is the expected operation; that works for me. That only leaves the issue with `realpath` failing during  the build (that's a problem that I didn't pick up with my review of #126903, but we can fix it here)

**Comment by hoodmane** [comment]
> the macOS version of realpath apparently doesn't honor the -s flag

Apple's crappy gnu forks strike again.

File: Tools/wasm/emscripten/web_example/python.worker.mjs
Comment (hoodmane) [reply Tools/wasm/emscripten/web_example/python.worker.mjs]: Well the way that I wanted to do this was blocked by
https://github.com/emscripten-core/emscripten/issues/22980
and I was waiting to see if it got fixed quickly or if I should find a workaround. Emscripten 3.1.73 was released with a fix so I think we can bump the minimum suggested Emscripten version to 3.1.73 and use the straightforward approach.

Comment (hoodmane) [reply Tools/wasm/emscripten/web_example/python.worker.mjs]: I suppose one simple fix would be to template the version into `python.sh` and then pass it in as an argument or environment variable. But I was thinking to get it from `Py_Version`.

**Comment by hoodmane** [comment]
> The Emscripten seems to work fine, however when viewing the web_example (using the provided server.py), I still get a CORS error:

@JeffersGlass This isn't a CORS error, you're serving it out of the source directory whereas you need to serve it out of the output directory:
```sh
./cross-build/wasm32-emscripten/web_example/server.py
```

File: Tools/wasm/README.md
Comment (freakboy3742) [inline Tools/wasm/README.md]:
```suggestion
To run the web example, ``cd`` into the ``web_example`` directory, then run 
``python server.py``. This will start a web server; you can then visit 
``http://localhost:8000/python.html`` in a browser to see a simple REPL example.
```

Also - I think this fits better above before "The web example uses SharedArrayBuffer...". The most important detail is the "prove it works" example usage; details about the specifics of the example are interesting once they've proved it works, and want to adapt it for their own usage.

File: Tools/wasm/emscripten/__main__.py
Comment (freakboy3742) [inline Tools/wasm/emscripten/__main__.py]: The `\` here leads to some odd indentation of the first comment line.
```suggestion
            f"""
```

File: Tools/wasm/emscripten/__main__.py
Comment (freakboy3742) [inline Tools/wasm/emscripten/__main__.py]: This is unnecessarily pejorative. macOS ships with the [FreeBSD version of coreutils](https://man.freebsd.org/cgi/man.cgi?query=realpath&sektion=1). While I agree the discrepancies between GNU and FreeBSD are often annoying, and in this instance the GNU implementation has useful features, that doesn't make the FreeBSD version a "defective fork".

File: Tools/wasm/emscripten/__main__.py
Comment (freakboy3742) [inline Tools/wasm/emscripten/__main__.py]: `--version` isn't a valid option for FreeBSD's realpath, so this generates output on stderr.
```suggestion
            elif which realpath > /dev/null && realpath --version 2&>1 | grep GNU > /dev/null; then
```

File: Tools/wasm/README.md
Comment (freakboy3742) [inline Tools/wasm/README.md]:
```suggestion
If you want to run the normal Python CLI, you can use `python.sh`. It takes the
same options as the normal Python CLI entrypoint, though the REPL does not
```

File: Tools/wasm/emscripten/__main__.py
Comment (freakboy3742) [inline Tools/wasm/emscripten/__main__.py]: Should this be:
```suggestion
                    echo "$(cd $(dirname "$1");pwd)/$(basename "$1")"
```
?

However, even then, if I try to run a `hello.py` (that contains `print("hello, world")`) that is in the root directory of the project, I get:
```
(venv) rkm@eunectes wasm32-emscripten % ./python.sh hello.py
realpath: illegal option -- -
usage: realpath [-q] [path ...]
/Users/rkm/projects/python/cpython/cross-build/wasm32-emscripten/python.sh: can't open file '//hello.py': [Errno 44] No such file or directory
```
Am I misunderstanding the intended usage?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
A couple of documentation tweaks; I'm still having issues with realpath as well. Not sure if I'm doing something wrong, though.

**Comment by hoodmane** [comment]
Thanks for the review @freakboy3742! I continue to appreciate your detailed PR feedback. I've been enjoying working with you so far on this.

File: Tools/wasm/emscripten/__main__.py
Comment (hoodmane) [reply Tools/wasm/emscripten/__main__.py]: Yeah... This is currently the expected output because of other bugs ("limitations") in `node_entry.mjs`. I wasn't planning to fix them in this PR because it's not relevant to the description.

The problems are:
1. Working directory not mounted into emscripten file system so it can't see `some_file.py`
2. Working directory not set correctly in emscripten so even if it were mounted it would look in the wrong place
3. Environment variables not propagated so in particular `HOME` is weird

Before then it won't work all that well like the python cli but with these changes it works pretty okay. I'm still also waiting on https://github.com/emscripten-core/emscripten/pull/22925 to get merged so that named pipes won't cause chaos.

Maybe I should make a separate PR and move the bsd realpath fixes and these three fixes there.

File: Tools/wasm/README.md
Comment (hoodmane) [reply Tools/wasm/README.md]: Indeed, makes sense.

File: Tools/wasm/emscripten/web_example/python.worker.mjs
Comment (hoodmane) [reply Tools/wasm/emscripten/web_example/python.worker.mjs]: Okay I fixed this in 7c61ed16707e6128b352575d57b6d267538ef3bc.

**Comment by hoodmane** [comment]
@freakboy3742 I reverted changes to `__main__.py` and moved them over to https://github.com/python/cpython/pull/127506.
I have made the requested changes; please review again (at least I think I made them...)

File: Tools/wasm/emscripten/__main__.py
Comment (hoodmane) [reply Tools/wasm/emscripten/__main__.py]: https://github.com/python/cpython/pull/127506

**Review by freakboy3742 — APPROVED** [review]
This obviously needs the changes from #127506 to be fully usable on macOS; but I'm happy with this as a set of changes in itself.

**Pull request merged**
